### ISSUE TITLE
Do not use "figure" tags for all highlighted code

### ISSUE THREAD

**Issue by stomar** [jekyll/jekyll#4905]
With #3779 the code wrapper was changed from `<div>` to `<figure>`. I consider this a strange decision and a bug.

In tutorials and blog posts it is very common to have small code snippets, like this:

``` sh
$ jekyll build
```

These code examples are usually highly interleaved with the surrounding text, very much like inline code spans.

Using a `<figure>` tag is semantically plainly **wrong**; the figure tag is meant for self-contained content that can be moved freely on the "page", like images, tables, figures (possibly with a caption) that are referenced from the text ("see fig. 2", ...). Of course, also code listings _can_ be figures, but only if the author wants them to be.

I understand the urge to remove `<div>` tags from html5 pages - I do not like them much myself - but thoughtlessly abusing a tag with a completely different purpose as a wrapper for every code block is a change for the worse.

Actually, I do not see the need for having an additional wrapper for highlighted code in the first place.  Currently all highlighted code is wrapped in a `<pre>`, a `<code>`, and  additionally a `<figure>` tag. But highlighted code is simply (preformatted) _code_, so why introduce an additional wrapper?

Forcing these wrongly placed figure elements upon the Jekyll user makes it more difficult to properly use figure elements. It really should be the choice of the author whether he wants his code examples to be figures or not.

**Comment by stomar** [comment]
**TL;DR**
Code blocks _can_ be figures, but they _do not have to be_, so they should not be wrapped in a `<figure>` tag by default.

**Comment by envygeeks** [comment]
It is your choice not to use `{% highlight %}` what's the problem?

**Comment by stomar** [comment]
???

I use `{% highlight %}` because I want to have highlighted code... that doesn't mean I want a figure; both are unrelated concepts.

**Comment by envygeeks** [comment]
@stomar it does, if the highlight tag chooses to use `<figure>` it's because the choice was made that it mostly applies to posts using code and relates to said cases.  You said however you did not have a choice in the matter and you do, if you do not wish for `<figure>` use <code>```</code>

I see no problem with using figure, you say it's wrong, but you do not say what is wrong with it in the context of what I just said we expect the `{% highlight %}` tag to be used.

**Comment by stomar** [comment]
I want **highlighted** code blocks, therefore I use `{% highlight %}` and not <code>```</code>. Since when does <code>{% highlight %}</code> mean "a figure containing highlighted code"? At least it hasn't in prior versions of Jekyll.

And in my original post I explained in detail why not every highlighted code is a figure.

**Comment by envygeeks** [comment]
@stomar AFAIW it was suggested at v3.0 which would make it about a year now it's existed. And <code>```</code> does highlight, as a matter of factly, it's the same highlighter we use for `{% highlight %}` by default (Rouge) now that we default to Kramdown.  The only difference here is that `{% highlight %}` takes the assumption it's used within a post and thus uses figure, <code>```</code> uses a div wrapper.

On my own sites I choose not to use code highlighting and highlight it with JS but you can infact use Rouge (or if it really pleases you Pygments.) However, by default, if you set no highlighter, we will default it to Rouge for you so that <code>```</code> works as expected, and if it doesn't, please do let us know what your configuration is (or if you can provide a base app that demonstrates it doesn't work) and it can be easily fixed (however, it should work by default as stated.)

**Comment by stomar** [comment]
It doesn't make a difference if it's used "within a post", and I really do not see why it "thus" (???) should use figures. The argument and the purpose of figure is the same for posts, pages, tutorials, books, whatever.

The point is that Jekyll provides a method for "Code snippet highlighting" (as described e.g. on https://jekyllrb.com/docs/templates/), and this method does assume more about the structure of the produced documents than it should. _You_ might want to always have your code snippets wrapped in a figure tag, but you shouldn't assume that's the only or best way to do it, generally.

BTW. Suggesting to use other means of code highlighting shouldn't be the point here; I am talking here about a functionality that Jekyll provides; your suggestions sound rather like "Jekyll is broken, so use something else", which is really not helpful.

**Comment by parkr** [comment]
@stomar We have been using `figure` for too long for us to remove it – I could see it breaking many sites. We did a `div` and someone corrected us to use a `figure`. Now you're saying we shouldn't use `figure`. If this didn't have the potential to break tons of sites, I'd definitely remove the `figure` because we shouldn't inject code really anyway, but we have to support it indefinitely. The triple backticks don't use `figure` for Rouge + kramdown (and possibly others) so feel free to move in that direction.

Thanks for making the suggestion though; we can always stand to improve as long as the cost isn't too high. 😄

**Comment by stomar** [comment]
Always sad to hear "it's too late to roll back"...

Suggestion for Jekyll 4:

> Let `{% highlight %}` behave in the most generic way, namely **only** producing `pre` and `code` tags, and not do any assumptions about whether the code snippet is a figure or not.

BTW. This had already been brought up by @kleinfreund several times in #3738.

_Regarding triple backticks: I explicitly deactivated kramdown's syntax highlighting, because (1) it complicates things to have highlighted code both from the liquid tag and from kramdown (different CSS classes and tag structure), and (2) there is a bug with kramdown's syntax highlighting that produces lots of inline code spans with class "err" (#4627)._

**Comment by envygeeks** [comment]
I see no problem with making it accept div or figure (by default) in Jekyll 4.  /cc @jekyll/stability @jekyll/ecosystem

**Comment by stomar** [comment]
What's wrong with completely dropping the additional wrapper and applying the CSS class directly to e.g. the `<code>` tag? There hasn't been a single argument against this possibility yet.

**Comment by spudowiar** [comment]
@stomar A lot of sites are currently relying on that `<figure>`

**Comment by stomar** [comment]
1. Are they? / Which ones? (We are only using `.highlight`)
2. Therefore Jekyll _4_.

I still would be interested in a _real_ argument for having this additional, semantically wrong tag.

**Comment by envygeeks** [comment]
This issue is being locked before things get too heated, what was needed to be said has been said and arguments are going towards the way of being personal and argumentative to a point where they are no longer constructive.
